cli/cook/querying.py

```diff
@@ -206,3 +216,57 @@ def query_unique_and_run(clusters, job_or_instance_uuid, command_fn):
         # This should not happen, because query_unique should
         # only return a map with type "job" or type "instance"
         raise Exception(f'Encountered error when querying for {job_or_instance_uuid}.')
+
+
+def query_with_pipe_support(clusters, uuids, command, pred_jobs=None, pred_instances=None,
+                            pred_groups=None, timeout=None, interval=None):
+    """
+    Queries for UUIDs across clusters, supporting input being passed via a pipe from another command, e.g.:
+
+      $ cs list --user sally --running --waiting | cs wait
+
+    The above example would wait for all of sally's running and waiting jobs to complete.
+    """
+    stdin_from_pipe = not sys.stdin.isatty()
+
+    if uuids and stdin_from_pipe:
+        raise Exception(f'When piping to {command}, you cannot also supply UUIDs as arguments.')
+
+    include_jobs = True
+    include_instances = True
+    include_groups = True
+
+    if stdin_from_pipe:
+        s = sys.stdin.read()
+        if not s:
+            raise Exception('No data received.')
+
+        try:
+            piped_data = json.loads(s)
+        except:
+            raise Exception('Unable to parse data.')
+
+        piped_cluster_names = set()
+        piped_types = set()
+        uuids = []
+        for datum in piped_data:
```

Comment:
Should we assert that `piped_data` is a `list` (or isn't a `dict`)?

Reply:
These comments about JSON might be irrelevant if you decide to switch to the format we were discussing with Steven yesterday afternoon, e.g.:
<i>public-cloud:group/13d2e8da-83aa-4b0c-a2be-6c86116db923</i>

Reply:
This now uses the URL format, e.g. `http://cook.example.com/jobs/13d2e8da-83aa-4b0c-a2be-6c86116db923`